WIP Temp PR

jamesfredley opened the pull request:
Key Changes and issues:

Groovy-5106 issues:

- `The interface GormEntity cannot be implemented more than once with different arguments: org.grails.datastore.gorm.GormEntity<grails.gorm.tests.XXX> and org.grails.datastore.gorm.GormEntity<grails.gorm.tests.XXX>`
- https://github.com/jamesfredley/grails-data-mapping/pull/1/files#diff-27ed87cb0bd085350a84cfb91c31242c7d3d516c64915076f927f2c746b25697
- https://github.com/jamesfredley/grails-data-mapping/pull/1/files#diff-e67a9d7f3d3f5d5c003fa33e3c6f4b8d65ecb714712905e92422fec02e4b2ccf
- https://github.com/jamesfredley/grails-data-mapping/pull/1/files#diff-4b36d1f95600c55e2dd8bfad1da2563ada49f20f037d441b40fc6639eee6c604
- https://github.com/jamesfredley/grails-data-mapping/pull/1/files#diff-9117157b9f4726fc00e80fbe03952dd5a9cdd9bf41ce93b9542fe38f7430309c
- https://github.com/jamesfredley/grails-data-mapping/pull/1/files#diff-eb7836a49960ae16314b800bd5db04b4d97950a933ceb9910f48707ea4b5ad12
- https://github.com/jamesfredley/grails-data-mapping/pull/1/files#diff-f7129bc6cdd92838d45ac5d0b544dce61291ab65cdb1fbcedaa9be74b9b05fad

Tenants changes - 
```
error: name clash: currentId(Class<? extends RxDatastoreClient>) in grails.gorm.rx.multitenancy.Tenants and currentId(Class<? extends Datastore>) in grails.gorm.multitenancy.Tenants have the same erasure, yet neither hides the other
public static  java.io.Serializable currentId(java.lang.Class<? extends org.grails.datastore.rx.RxDatastoreClient> datastoreClass) { return (java.io.Serializable)null;}
```
```
Tenants.java:11: error: name clash: <T#1>withCurrent(Class<? extends RxDatastoreClient>,Closure<T#1>) in grails.gorm.rx.multitenancy.Tenants and <T#2>withCurrent(Class<? extends Datastore>,Closure<T#2>) in grails.gorm.multitenancy.Tenants have the same erasure, yet neither hides the other
public static <T> T withCurrent(java.lang.Class<? extends org.grails.datastore.rx.RxDatastoreClient> datastoreClass, @groovy.lang.DelegatesTo(value=org.grails.datastore.rx.RxDatastoreClient.class) groovy.lang.Closure<T> callable) { return (T)null;}
                    ^
  where T#1,T#2 are type-variables:
    T#1 extends Object declared in method <T#1>withCurrent(Class<? extends RxDatastoreClient>,Closure<T#1>)
    T#2 extends Object declared in method <T#2>withCurrent(Class<? extends Datastore>,Closure<T#2>)
```
- https://github.com/jamesfredley/grails-data-mapping/pull/1/files#diff-7708c0f3847869b8ff4971104b44c9747b78707413ab99da13e34981f7e21691

Annotation Metadata Reader

- https://github.com/jamesfredley/grails-data-mapping/pull/1/files#diff-c766af43d516aa0d817579bbd8ec67e3fca6d2ef567f172b415fe88f7901650e

javax in Grails Package hierarchy and class name

- https://github.com/jamesfredley/grails-data-mapping/pull/1/files#diff-32943b7b556a0f60770ec10dada5053c28e619dc6d88b955decc7e018d69db70
- renamed package and class

CastExpression.asExpression(newReturnType, returnExpr)

- https://github.com/jamesfredley/grails-data-mapping/pull/1/files#diff-298a9c9326beaac1ddf877eb26c396572831812cf35d52667e1cf8e480b76297

`The interface XXX cannot be implemented more than once with different arguments: java.util.Example<java.lang.Object> and java.util.Example<E>`

- https://github.com/jamesfredley/grails-data-mapping/pull/1/files#diff-e77ee4b62b7ff830feddde3a0c6d2c3b073927f75e70a1b0d43c527136c76d80
- https://github.com/jamesfredley/grails-data-mapping/pull/1/files#diff-01a7e593dc1261f2ab713f5e740bdcf8b222d2f4ec935fa70f8bb97f45aace68
- https://github.com/jamesfredley/grails-data-mapping/pull/1/files#diff-fa2aeeaeb7a85f0991bf9319dc119a425a32597fa99ac61a374452863416dcf1
- The changes to extends and implements on RxPersistentSortedSet.groovy resolved the above rather generic error: `org.codehaus.groovy.syntax.RuntimeParserException: The interface Set cannot be implemented more than once with different arguments: java.util.Set<java.lang.Object> and java.util.Set<E>`
- https://github.com/jamesfredley/grails-data-mapping/pull/1/files#diff-0717c9ff3dac125db0a958e2f0ae1b845bf6e703957e29148b2aaedcc36e8fc0
- https://github.com/jamesfredley/grails-data-mapping/pull/1/files#diff-d3b653109b246c87c34c3e2ab9aefa6d716b2d9297d9048ec81108da291911ee
- https://github.com/jamesfredley/grails-data-mapping/pull/1/files#diff-d7496229b30eb8468fc3bb6f3805ad6e3233a8a74e1150723255ea2682fe113a
- https://github.com/jamesfredley/grails-data-mapping/pull/1/files#diff-14b92adaede32ac0daf85e12325a785aa7738c662ac47967fdb1f0ec1b061bb8


More test failures:
- https://github.com/jamesfredley/grails-data-mapping/pull/1/files#diff-50a99d77e04ca2ce8390a54c4b292ecd43315ce23a39cba456d20b2438f0e0e5
`Cannot implement method for argument [title], No implementations possible for method 'rx.Observable updateBook(java.lang.String, java.lang.String)`
- https://github.com/jamesfredley/grails-data-mapping/pull/1/files#diff-91b8d65fd4ce06aba9ae88d79069c79c84cccd1bf920d78579fb79ec4058792f
- `java.lang.IllegalStateException: Either class [grails.gorm.services.multitenancy.partitioned.Book] is not a domain class or GORM has not been initialized correctly or has already been shutdown. Ensure GORM is loaded and configured correctly before calling any methods on a GORM entity.`
- https://github.com/jamesfredley/grails-data-mapping/pull/1/files#diff-5308fcc21bbd00b74eb7fe0fb1b844293ea4be3c3e22be1b0f1858ece8144a85
`java.lang.IllegalStateException: Either class [grails.gorm.services.multitenancy.schema.Book] is not a domain class or GORM has not been initialized correctly or has already been shutdown. Ensure GORM is loaded and configured correctly before calling any methods on a GORM entity.`
-https://github.com/jamesfredley/grails-data-mapping/pull/1/files#diff-db6628bfbaa9316ae15a143760a4e8f08ab13d7b3b9768f1889f87c4c2303d18
`java.lang.StackOverflowError`

matrei committed:
Update h2 test dependency to 2.2.224 (#1768)

renovate[bot] committed:
Update spring core to v5.3.31 (#1769)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

puneetbehl committed:
[skip ci] Release v

puneetbehl committed:
[skip ci] back to snapshot

puneetbehl committed:
[skip ci] Release v8.0.3

puneetbehl committed:
Back to 8.0.4-SNAPSHOT

renovate[bot] committed:
Update plugin com.gradle.common-custom-user-data-gradle-plugin to v1.12.1 (#1771)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update dependency commons-validator:commons-validator to v1.8.0 (#1772)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

puneetbehl committed:
Update GitHub Workflows (#1773)

renovate[bot] committed:
Update plugin com.gradle.enterprise to v3.16 (#1770)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update plugin com.gradle.enterprise to v3.16.1 (#1775)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update grails monorepo to v6.1.1 (#1774)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

matrei committed:
docs: Update readme (#1777)

* docs: Update readme

- change to https links
- make links inline
- update md syntax
- add maven central badge
- add event filter (push) to CI badge
- remove release workflow badge
- improve text

* docs: fix maven central badge url

renovate[bot] committed:
Update dependency org.javassist:javassist to v3.30.2-GA (#1776)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

puneetbehl committed:
Add Groovy joint workflow (#1778)

* Add groovy-joint-workflow

* Update groovy-joint-workflow.yml

matrei committed:
fix: Groovy 3.0.20 compatibility (#1779)

puneetbehl committed:
Override project groovyVersion from env CI_GROOVY_VERSION (#1780)

* Update build.gradle

Override project groovyVersion from env CI_GROOVY_VERSION

* Force groovy version from the env CI_GROOVY_VERSION

puneetbehl committed:
Only publish build scan for CI (#1781)

puneetbehl committed:
[skip ci] Release v8.0.4

puneetbehl committed:
Back to 8.0.5-SNAPSHOT

guillermocalvo committed:
Fix initialization of `ValidatedService.validatorFactory` (#1784)

* Fix trait `ValidatedService` (broken `validatorFactory` initialization)

* Annotate instance field initialization with `@CompileDynamic`

* Drop `@CompileDynamic`; rename instance field to avoid name collision

renovate[bot] committed:
Update grails monorepo to v6.1.2 (#1783)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

puneetbehl committed:
Update to 8.1.0-SNAPSHOT

renovate[bot] committed:
Update plugin com.gradle.common-custom-user-data-gradle-plugin to v1.12.2 (#1787)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update dependency gradle to v7.6.4 (#1785)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update groovy monorepo to v3.0.20 (#1731)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update spring core to v5.3.32 (#1786)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update plugin com.gradle.enterprise to v3.16.2 (#1782)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

puneetbehl committed:
[skip ci] Release v8.1.0

puneetbehl committed:
Back to 8.1.1-SNAPSHOT

renovate[bot] committed:
chore(deps): update plugin com.gradle.common-custom-user-data-gradle-plugin to v1.13 (#1789)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
fix(deps): update groovy monorepo to v3.0.21 (#1790)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

jhron committed:
Added TenantDataSourceConfig annotation to add possibility exclude datasource to be used by multi-tenancy. (#1380)

puneetbehl committed:
[skip ci] Release v8.1.1

puneetbehl committed:
Back to 8.1.2-SNAPSHOT

puneetbehl committed:
chore(release): enhance release workflow configurations (#1791)

* Break down into multiple jobs
* Fix publishing of docs

puneetbehl committed:
chore(build): Migrate to Develocity build cache connector (#1792)

To simplify cache management with  the [Develocity cache connector](https://docs.gradle.com/enterprise/gradle-plugin/#using_the_develocity_connector)

puneetbehl committed:
chore(releasenotes): update workflow configuration (#1794)

* Improve `release-drafter.yml` configuration template file.
* Add `pull_request` event is required only for autolabeler.
* Add `pull_request_target` event is required for autolabeler to support PRs from forks.

renovate[bot] committed:
fix(deps): update spring core to v5.3.33 (#1793)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

puneetbehl committed:
Update release-drafter.yml (#1796)

puneetbehl committed:
chore(release): Update workflow configuration (#1795)

Fix incorrect indentation in the `release.yml` workflow file.

puneetbehl committed:
chore(ci): improve release-drafter configuration (#1797)

Remove labels, "type: dependency upgrade", deps, and dependencies from the Maintenance category.

puneetbehl committed:
[skip ci] Release v8.1.2

puneetbehl committed:
chore: Bump version to 8.1.3-SNAPSHOT

puneetbehl committed:
chore(release): update workflow configurations (#1798)

guillermocalvo committed:
Migrate to Develocity build cache connector (#1799)

renovate[bot] committed:
fix(deps): update grails monorepo to v6.2.0 (#1801)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update plugin com.gradle.enterprise to v3.17.1 (#1800)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

jamesfredley committed:
Merge branch '9.0.x' into jamesfredley/9.0.x-changes

# Conflicts:
#	gradle.properties

jamesfredley committed:
WIP

jamesfredley committed:
WIP 2

jamesfredley committed:
Update gradle properties

jamesfredley committed:
WIP 3

jamesfredley committed:
Add execption from Groovy 5106 errors

sbglasius commented on grails-datastore-gorm/src/main/groovy/grails/gorm/multitenancy/Tenants.groovy:
Any reason why the Generics on the `Closure callable` parameter was dropped. That would be the return value type (`<T>`)

sbglasius reviewed (commented):
I just read through the changes, and I had one comment. I don't know the inner gists of GORM, so please take my review with a grains of salt.

jdaugherty commented on grails-datastore-gorm/src/main/groovy/grails/gorm/multitenancy/Tenants.groovy:
I believe this was because there's some odd type erasure going on with groovy 4: 

    grails-datastore-gorm-rx\build\tmp\compileGroovy\groovy-java-stubs\grails\gorm\rx\multitenancy\Tenants.java:11: error: name clash: <T#1>withCurrent(Class<? extends RxDatastoreClient>,Closure<T#1>) in grails.gorm.rx.multitenancy.Tenants and <T#2>withCurrent(Class<? extends Datastore>,Closure<T#2>) in grails.gorm.multitenancy.Tenants have the same erasure, yet neither hides the other
    public static <T> T withCurrent(java.lang.Class<? extends org.grails.datastore.rx.RxDatastoreClient> datastoreClass, @groovy.lang.DelegatesTo(value=org.grails.datastore.rx.RxDatastoreClient.class) groovy.lang.Closure<T> callable) { return (T)null;}
                        ^
      where T#1,T#2 are type-variables:
        T#1 extends Object declared in method <T#1>withCurrent(Class<? extends RxDatastoreClient>,Closure<T#1>)
        T#2 extends Object declared in method <T#2>withCurrent(Class<? extends Datastore>,Closure<T#2>)
    Note: Some input files use unchecked or unsafe operations.
    Note: Recompile with -Xlint:unchecked for details.

jdaugherty commented:
We were discussing the grails-data-mapping upgrade for grails 7.  So far most people have commented out code that is affected by https://issues.apache.org/jira/browse/GROOVY-5106 .  This bug is really a blocker - it prevents inheritance in GORM.  Inheritance is a feature of hibernate and we need to continue to support it.  Here's some impacted code (you can run just groovyc on this code):

      trait GormEntity<D> {
        D getAll() {
          null
        }
      }

      class Parent implements GormEntity<Parent> {

      }


      class Child extends Parent implements GormEntity<Child> {

      }

Trying to compile this code in groovy 4 would give you this error:

`The interface GormEntity cannot be implemented more than once with different arguments: GormEntity<Child> and GormEntity<Parent>`

The groovy ticket prevents having a Parent domain & Child domain, where you want to be able to query both the Parent or the Children.  We need to question the original groovy change: why do generic types of traits not follow the same rules for multiple inheritance in groovy?  That is, the generic type distinguishes between the same trait class.  

As a further reason this should be true, the following is allowed in groovy:

      trait GormEntityParent {
        Parent getAll() {
          null
        }
      }

      trait GormEntityChild {
        Child getAll() {
          null 
        }
      }


      class Parent implements GormEntityParent {
        
      }

      class Child extends Parent implements GormEntityChild {
        
      }

GORM is just wanting to implement this same example with generics.  @jamesfredley pointed out that this validation was added under a single revertable commit in groovy.  If reverted, we would be able to move forward.

jamesfredley commented on grails-datastore-gorm/src/main/groovy/grails/gorm/multitenancy/Tenants.groovy:
It was one of many attempts to work through issues with Tenants.groovy and Groovy 4.0.22  I think that class needs attacked from scratch by someone more knowledgeable about the erasure issue with Closures as a method parameter.

jamesfredley committed:
Rename org.grails.datastore.gorm.validation.javax to org.grails.datastore.gorm.validation.javax
and JavaxValidatorRegistry.groovy to JakartaValidatorRegistry.groovy

jamesfredley committed:
Cleanup name clash between methods that have the same erasure, yet neither hides the other between
grails.gorm.multitenancy.Tenant and grails.gorm.rx.multitenancy.Tenant by removing inheritance

jamesfredley committed:
Clean up tests

jamesfredley committed:
More test cleanup

jamesfredley commented:
@jdaugherty @sbglasius 

I was able to find a truly simple solution to cleanup the name clash between methods that have the same erasure, yet neither hides the other between
`grails.gorm.multitenancy.Tenant` and `grails.gorm.rx.multitenancy.Tenant` by removing the unnecessary inheritance between them

I also searched and found that the following change was limited to grails-data-mapping and gorm-hibernate5, so I proceeded with it.
Rename `org.grails.datastore.gorm.validation.javax` to `org.grails.datastore.gorm.validation.jakarta`
and `JavaxValidatorRegistry.groovy` to `JakartaValidatorRegistry.groovy`

And lastly I went through all of the prior failing tests and was able to remove 2 that are now passing and regrouped one that had nothing to do with domain inheritance.  

So this might be down to just failing tests at this point.  

6 for domain inheritance
4 others

jamesfredley committed:
Java 17 updates on github workflows

jamesfredley committed:
More github workflow cleanup

jamesfredley committed:
Update github workflow

jamesfredley committed:
Update copyright

jamesfredley commented:
was completed on https://github.com/grails/grails-data-mapping/pull/1810
